web/src/saved_snippets_ui.ts

```diff
@@ -0,0 +1,110 @@
+import type {StateData} from "./state_data";
+
+let saved_snippet_dropdown_widget: dropdown_widget.DropdownWidget;
+
+function submit_create_saved_snippet_form(): void {
+    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .title").val()?.trim();
+    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .content").val()?.trim();
+    if (title && content) {
+        dialog_widget.submit_api_request(channel.post, "/json/saved_snippet", {title, content});
+    }
+}
+
+function update_submit_button_state(): void {
+    const title = $<HTMLInputElement>("#add-new-saved-snippet-modal .title").val()?.trim();
+    const content = $<HTMLInputElement>("#add-new-saved-snippet-modal .content").val()?.trim();
+    const $submit_button = $("#add-new-saved-snippet-modal .dialog_submit_button");
+
+    $submit_button.prop("disabled", true);
+    if (title && content) {
+        $submit_button.prop("disabled", false);
+    }
+}
+
+function saved_snippet_modal_post_render(): void {
+    $("#add-new-saved-snippet-modal").on("input", "input,textarea", update_submit_button_state);
+}
+
+export function rerender_dropdown_widget(): void {
+    const options = saved_snippets.get_options_for_dropdown_widget();
+    saved_snippet_dropdown_widget.list_widget?.replace_list_data(options);
+}
+
+function delete_saved_snippet(saved_snippet_id: string | undefined): void {
+    assert(saved_snippet_id !== undefined);
+    void channel.del({
+        url: "/json/saved_snippet/" + saved_snippet_id,
+    });
+}
+
+function item_click_callback(
+    event: JQuery.ClickEvent,
+    dropdown: tippy.Instance,
+    widget: dropdown_widget.DropdownWidget,
+): void {
+    event.preventDefault();
+    event.stopPropagation();
+
+    if (
+        $(event.target).closest(".saved_snippets-dropdown-list-container .dropdown-list-delete")
+            .length
+    ) {
+        delete_saved_snippet($(event.currentTarget).attr("data-unique-id"));
+        return;
+    }
+
+    dropdown.hide();
+    const current_value = widget.current_value;
+    if (current_value === saved_snippets.ADD_SAVED_SNIPPET_OPTION_ID) {
+        dialog_widget.launch({
+            html_heading: $t_html({defaultMessage: "Add a new saved snippet"}),
+            html_body: render_add_saved_snippet_modal(),
+            html_submit_button: $t_html({defaultMessage: "Save"}),
+            id: "add-new-saved-snippet-modal",
+            form_id: "add-new-saved-snippet-form",
+            update_submit_disabled_state_on_change: true,
+            on_click: submit_create_saved_snippet_form,
+            post_render: saved_snippet_modal_post_render,
+        });
+    } else {
+        const content = saved_snippets
+            .get_saved_snippets()
+            .find((saved_snippet) => saved_snippet.id === current_value)?.content;
+        assert(content !== undefined);
+        const $textarea = $<HTMLTextAreaElement>("textarea#compose-textarea");
```

Comment:
When we make these available in the message-edit form as well, we'll need to make this a parameter to this function. Might be worth doing that plumbing now just to avoid having to do more refactoring later, but it's not strictly required.

Reply:
Did you mean to say that I should make `saved_snippet_id` a parameter to `get_saved_snippets`?

Reply:
Yeah.